File: JavaScript/Arrays/ArrayedSet/ArrayedSet.js
/*******************************************************
 *  ArrayedSet.js
 *  Created by Stephen Hall on 11/15/17.
 *  Copyright (c) 2017 Stephen Hall. All rights reserved.
 *  Arrayed Set implementation in JavaScript
 ********************************************************/

/**
 * Arrayed Set object
 * @param size: size to initialize to
 * @constructor
 */
function ArrayedSet(size) {
    if(size !== undefined && size !== null && size > 0)
        this.size = size;
    else
        this.size = 4;

    this.array = [this.size];
    this.count = 0;

    function resize() {
        this.size *= 2;
    }
}

/**
 * Adds data into the array
 * @param data: data to add
 * @returns {*}; data added or null if already exists
 */
ArrayedSet.prototype.add = function(data) {
    if(data === null || data === undefined)
        return null;

    if(this.contains(data))
        return null;

    if(this.count === this.size)
        resize();

    this.array[this.count] = data;
    this.count++;
    return data;
};

/**
 * Appends the contents of an array to the ArrayedSet
 * @param data: array to append
 * @returns {boolean} success|fail
 */
ArrayedSet.prototype.append = function(data){
    if(data !== null){
        for (var i = 0; i < data.length; i++) {
            if(data[i] !== null )
                this.add(data[i]);
        }
        return true;
    }
    return false;
};

/**
 * Sets the data at the given index
 * @param index: index to set
 * @param data: data to set index to
 * @returns {boolean}: success|fail
 */
ArrayedSet.prototype.set = function(index, data) {
    if(!this.contains(data)){
        if(index >= 0 && index < this.size) {
            this.array[index] = data;
            return true;
        }
    }
    return false;
};

/**
 * Gets the data at the arrays given index
 * @param index: Index to get data at
 * @returns {*}: Data at the given index or default value of T if index does not exist
 */
ArrayedSet.prototype.get = function(index) {
    if(index >= 0 && index < this.size)
        return this.array[index];
    return null;
};

/**
 * Removes the data at arrays given index
 * @param index: Index to remove
 * @returns {*}: Data removed from the array or null
 */
ArrayedSet.prototype.remove = function(index) {
    if (index === undefined || index === null || index < 0 || index > this.count)
        return null;

    var tmp = this.array[index];
    this.array[index] = null;
    this.count--;
    return tmp;
};

/**
 * Resets the internal array to default size with no data
 */
ArrayedSet.prototype.reset = function() {
    this.count = 0;
    this.size = 4;
    this.array = [this.size];
};

/**
 * Clears all data in the array leaving size intact
 */
ArrayedSet.prototype.clear = function() {
    for(var i = 0; i < this.count; i++){
        this.array[i] = null;
    }
    this.count = 0;
};

/**
 * Tests to see if the data exist in the list
 * @param data: data to find
 * @returns {boolean}: true|false
 */
ArrayedSet.prototype.contains = function(data) {
    if(data !== undefined && data !== null)
    for(var i = 0; i < this.size; i++){
        if(this.array[i] === data)
            return true;
    }
    return false;
};

/**
 * Gets the current count of the array
 * @returns {ArrayedSet.count|*}: umber of items in the array
 */
ArrayedSet.prototype.count = function() {
    return this.count;
};

File: C++/Arrays/ArrayedSet/ArrayedSet.cpp
/*******************************************************
 *  @file ArrayedSet.cpp
 *  @author Stephen Hall
 *  @date 11/22/17.
 *  @copyright 2017 Stephen Hall. All rights reserved.
 *  @details ArrayedSet implementation in C++
 ********************************************************/

/* Only needed because I did not include a file where it is defined */
#ifndef NULL
#define NULL   ((void *) 0)
#endif

#include "ArrayedSet.h"
using namespace Arrays::ArrayedSet;

/**
 * ArrayedSet default constructor
 * @tparam T: Generic type
 */
template <typename T>
ArrayedSet::ArrayedSet()
{
	array = new T[size = 4];
	for(int i = 0; i < size; i++)
		array[i] = NULL;
	count = 0;
}

/**
 * ArrayedSet Constructor initialized to given size
 * @tparam T: Generic Type
 * @param size: Size to initialize the array list to
 */
template <typename T>
ArrayedSet::ArrayedSet(unsigned int size)
{
	array =  new T[this->size = size > 0 ? size : 4];
	for(int i = 0; i < size; i++)
		array[i] = NULL;
	count = 0;
}

/**
 * Adds data into the ArrayedSet
 * @tparam T: Generic Type
 * @param data: data to add into the ArrayedSet
 * @return T: Data added into the ArrayedSet
 */
template <typename T>
T ArrayedSet::add(T data)
{
	if(data == NULL || contains(data))
		return NULL;

	if(count == size)
		resize();

	array[count] = data;
	count++;

	return data;
}

/**
 * Appends an Array of T onto the ArrayedSet
 * @tparam T: Generic Type
 * @param dataArray: array to append
 * @param size: size of the array
 * @return bool: success|fail
 */
template <typename T>
bool ArrayedSet::append(T* dataArray, int size)
{
	if(dataArray != NULL){
		for(int i = 0; i < size; i++)
		{
			if (dataArray[i] != NULL)
				add(dataArray[i]);
		}
		return true;
	}
	return false;
}

/**
 * Clears all the elements in the Array List but keeps the size intact
 * @tparam T: Generic Type
 */
template <typename T>
void ArrayedSet::clear()
{
	for(int i = 0; i < size; i++)
		array[i] = NULL;
}

/**
 * ArrayedSet Destructor
 * @tparam T: Generic Type
 */
template <typename T>
ArrayedSet::~ArrayedSet()
{
	delete [] array;
	array = NULL;
}

/**
 * Gets the number of elements in the ArrayedSet
 * @return
 */
int ArrayedSet::length()
{
	return count;
}

/**
 * Overloaded [] operator for array indexing
 * @tparam T: Generic type
 * @param index: index to access
 * @return T: data at the given index or null
 */
template <typename T>
T& ArrayedSet::operator[](int index)
{
	return (index < size) ? array[index] : NULL;
}

/**
 * Removes the first instance of the given data from the ArrayedSet
 * @tparam T: generic type
 * @param data: Data to remove
 * @return T: data removed from the ArrayedSet
 */
template <typename T>
T ArrayedSet::remove(T data)
{
	for(int i = 0; i < count; i++)
	{
		if(array[i] == data)
		{
			T tmp = array[i];
			array[i] = NULL;
			return tmp;
		}
	}
	return NULL;
}

/**
 * Resets the array to its default state
 * @tparam T: Generic Type
 */
template <typename T>
void ArrayedSet::reset()
{
	delete [] array;
	array = new T[size = 4];
	for(int i = 0; i < size; i++)
		array[i] = NULL;
	count = 0;
}

/**
 * Resize's the ArrayedSets internal array
 * @tparam T: Generic Type
 */
template <typename T>
void ArrayedSet::resize()
{
	size *= 2;
	T* tmp =  new T[size];
	for(int i = 0; i < length(); i++)
		tmp[i] = array[i];

	delete [] array;
	array = tmp;
	array = tmp;
}

/**
 * Determines if the data is contained in the ArrayedSet
 * @tparam T: Generic Type
 * @param data: data to test for
 * @return bool: yes|no
 */
template <typename T>
bool ArrayedSet::contains(T data)
{
	if(data != NULL)
		for(int i = 0; i < count; i++)
			if(array[i] == data)
				return true;
	return false;
}